### ISSUE TITLE
Use of optional vs required nullable attributes in schema

### ISSUE THREAD

**Issue by reece** [ga4gh/vrs#275]
The schema currently uses the `required` attribute inconsistently. It is unclear whether a not-required attribute is equivalent to a nullable attribute.

For this issue, investigate how to make the schema more consistent.

N.B. Reece has a vague recollection that python-jsonschema-objects conflates default values and nullable values during serialization.

**Comment by ahwagner** [comment]
To restate: we are debating whether we want to express nullable attributes as optional, or if they are each required but can accept `NULL` / `None` values.

This conversation is probably deeper than we want to get into ahead of v1.2 release, so we are setting aside as a separate concern.

**Comment by rrfreimuth** [comment]
I think the concepts are separate, so we should base the decision on what semantics we want to convey.  Since both stated options are capable of expressing similar semantics, the decision should be based on the _interpretation_ of what is expressed.  I favor explicit over implicit, so in cases where a missing attribute could be interpreted (or intended to be interpreted) as "null" then I would prefer to remove the optionality so differences in interpretation are minimized.

**Comment by reece** [comment]
I agree with Bob: I prefer the precision of an explicit null over an optional attribute.

**Comment by ahwagner** [comment]
I also agree. Now I'm wondering if we should use explicit NULL or consider other explicit values that hint at their meaning?

**Comment by larrybabb** [comment]
Food for thought https://www.wired.com/story/null-license-plate-landed-one-hacker-ticket-hell/

Just because I've never worked on a production system that intentionally loaded all optional attributes with a special string that meant "intentionally NO value" or "null" doesn't mean that this is wrong, but I am scratching my head to figure out how/why we are in this predicament.  Can we at least cite other reputable software or messaging specs that require the use of a value for attributes that are NULL? I'm sure I'm just missing it - but I can't recall any off the top of my head.

**Comment by larrybabb** [comment]
Another interesting article ... https://www.lucidchart.com/techblog/2015/08/31/the-worst-mistake-of-computer-science/

**Comment by ahwagner** [comment]
Yeah, I have the same reservations as @larrybabb about requiring `some_type` | `NULL` on attributes. But I also think an explicit, required value is a good thing.

**Comment by rrfreimuth** [comment]
Case 1: `{ attrib_1 = "A", attrib_2 = NULL }`
Case 2: `{ attrib_1 = "A" }`

Do these mean the same thing?  I think there is a subtle difference between them.  Case 1 means "the value of attrib_2 is explicitly NULL".  Case 2 could mean either "the value of attrib_2 is IMplicitly NULL" or "attrib_2 may or may not have a value; no assertion about the value of attrib_2 is made in this message".

My point was more about the inference of a value for an attribute that is not present than it was about the value that is used to indicate the concept of NULL.  FWIW, HL7 has a whole [code system](https://www.hl7.org/fhir/v3/NullFlavor/cs.html) to represent the latter.

**Comment by reece** [comment]
That's a chilling article. To be clear, we're not using *strings* that refer to null values (of any name), we're using null values themselves. The article exposes the problems that result from confusing named literals (like `NULL`) with string literals (like the string `"NULL"`).

> Can we at least cite other reputable software or messaging specs that require the use of a value for attributes that are NULL? 

All relational databases since 1970?

All of them have a notion of NULL values, and rows always include all specified columns, including those with null values. A required attribute that may be nullable is exactly the same semantics as a db column that is nullable.

What would seem weirder to me is for a database to return only the columns that are not null. Different records would have different composition. That would be especially problematic if you digested the record to create an identifier.

**Comment by reece** [comment]
@rrfreimuth wrote:

> Case 2 could mean either "the value of attrib_2 is IMplicitly NULL" or "attrib_2 may or may not have a value; no assertion about the value of attrib_2 is made in this message".

Actually, I think case 2 really says that attrib_2 has its default value, which might be null or something else. For example, if the default value is 0, then case 2 might be just omitting the default value.

The assumption that the default value is null is insidious and is the major reason I think we should always show attributes.

Consider this case: Class C has an attribute a with a default value of 5. Then, we create c1 = C(a=null). With explicit attributes, it serializes to `{a: null}`; with implicit null attributes, this serializes to `{}`. Now imagine a client interpreting these two messages. Rehydrating the first will give us C(a=null), which is correct; rehydrating the second will give C(a=5), which uses the default value.

**Upshot: Implicit null representations fail for any attribute that has a non-null default value.**

**Comment by rrfreimuth** [comment]
@reece I agree with you, but I believe the assumption that "not provided" means "default value" requires a bit of business logic. I think it is reasonable for applications to both use default values but not assert those values for attributes that are not provided.  De-identification of data sets is an example.  Just because a value has been stripped out of a particular message should not always be interpreted to mean that it has a default value (although it might be documented to be so in some cases).

Bottom line: I agree that implicit is undesirable, but not singularly because of the ambiguity regarding defaults.

**Comment by larrybabb** [comment]
If an attribute is not included in a message then it should not mean anything other than the attribute is not included. 

If someone wants to exchange a message structure that contains attributes that have no values then you would use the serialization format's method for conveying nulls. For example json spec has a `null` term you can use in a message for the value with no quotes, braces or brackets.

> [JSON empty string vs null](https://www.xspdf.com/resolution/52223257.html#:~:text=If%20you%20want%20to%20represent,in%20JSON%20data%20is%20described.)
> If you want to represent a null value in JSON, the entire JSON string (excluding the quotes containing the JSON string) is simply null. No braces, no brackets, no quotes.

[Here's](https://stackoverflow.com/questions/21120999/representing-null-in-json/21122019) a good stackoverflow thread on the topic with the clear winner being choice 2 (same as above)

Why would we get involved in defining a special value for null?  And if anyone uses our spec to pass partial lists of attributes I'm not sure we should make any explicit declaration about what the consumer of such a message should infer the value to be or not to be. It just isn't shared. If it is essential to the meaning of the concept then it would be an incomplete data type.

**Comment by larrybabb** [comment]
i'm of the opinion that if an optional attribute does not have a value it should not be required to send in the message. If it is truly optional then let's say that you should only pass it when it is not null. If it impacts the meaning of the object it is in, then why is it optional? Anything that is required will have non-null values.  Can we just wait until we hit a scenarios where we are required to deal with this at a spec level?

**Comment by reece** [comment]
There's so much commenting I can't tell what's actually being proposed.

A. Attributes should *always* be included in messages, regardless of value (incl. null)
B. Attributes should be included if they're not null
C. Attributes should be included if they're not the default value
D. Something else that I'm too dense to glean from the above convo.

Me: A.
Second choice: C, knowing that it's usually the same as case B because most defaults are null.

**Comment by rrfreimuth** [comment]
A.

**Comment by ahwagner** [comment]
@larrybabb 
> i'm of the opinion that if an optional attribute does not have a value it should not be required to send in the message. If it is truly optional then let's say that you should only pass it when it is not null. If it impacts the meaning of the object it is in, then why is it optional? Anything that is required will have non-null values. Can we just wait until we hit a scenarios where we are required to deal with this at a spec level?

We do need to address this for the current spec. I'm thinking about intervals here: one of start or end is required, but not both.

**Comment by larrybabb** [comment]
I think B is the right answer. But, why are we trying to dictate whether developers must include attributes at all? Imagine if I defined an Address concept with line1, line2, city, state, zip, country attributes and told folks that you had to always include every attribute - it doesn't make sense. If someone wants to only pass the city and state, so be it. Our job is to define the purpose of the attributes, demonstrate the intended use and then let folks explore and use them as they wish. If they use them improperly then that's on them. Maybe they'll even find creative ways to apply them that we have not considered.

The default value thing is a non-starter for me. I just don't understand the logic of presuming everyone will know the default value and build that into their systems. There should be no (or very little) assumptions about what the sender intended to send. 

If someone wants to explicitly send an attribute with an unassigned value then I would defer to the serialization mechanism being used by the implementers (null for json and xsi.nil=true attribute for XML elements). I don't think we want to get into the rats nest of trying to define nullness.

**Comment by larrybabb** [comment]
changed to stayin alive since we are trying to address this uniformly starting with release 1.2RC0.

**Comment by ahwagner** [comment]
I agree with @larrybabb that if we are not creating explicit negative attributes (e.g. use of a comparator string for describing unbounded ranges) but are instead using absence of value to have meaning (e.g. required use of only one of `min` or `max` for `Range` in #286), then we should not create requirements for an explicit `null`. JSON Schema [interprets null to mean "a missing value"](https://json-schema.org/understanding-json-schema/reference/null.html), though there is a [clear (and logical) difference in validator behavior](https://www.jsonschemavalidator.net/s/c0HMWYfO) for optional attributes and required nullable attributes.

Using optionality instead of explicit `null` is also the more lightweight choice, since fields do not need to be polymorphic–this has consequences when trying to represent VRS in other systems / technologies (e.g. protocol buffers, which we are adding support for in the [y2p branch](https://github.com/ga4gh/vrs/tree/y2p)). My current, initial implementation in #286 is to allow optional `min` and `max`.

Also, I'm going to change the title of this issue to focus specifically on optional vs. explicit Null attributes, as I think that the use of `required` across classes in JSON Schema is a related, but separate issue.

**Comment by ahwagner** [comment]
After further discussion, @reece, @larrybabb, and I have recognized the need for an explicit and non-optional representation of _all_ attributes in VRS, except for those beginning with an underscore (i.e., `_id`).

This will necessitate additional changes to the spec ahead of v1.2, reverting to a Number / ComparatorNumber / Range class set (vs. the current multi-purpose Range class).
